src/append/dedup.rs

```diff
@@ -0,0 +1,94 @@
+//! The dedup common handler
+//!
+use crate::append::Error;
+use crate::encode::{Encode, Write};
+use log::Record;
+
+const REPEAT_COUNT: i32 = 1000;
+/// dedup object to be used by deduping appender.
+/// internals are private to dedup
+#[derive(Default)]
+pub struct DeDuper {
+    count: i32,
+    last: String,
+}
+#[derive(PartialEq)]
+/// Used by an appender that uses dedup.
+/// Indicates whether or not the current message should be output.
+///
+/// sample use from console appender
+///         if let Some(dd) = &self.deduper {
+///              if dd.lock().dedup(&mut *file, &*self.encoder, record)? == DedupResult::Skip {
+///                  return Ok(());
+///             }
+///     ... output the message
+pub enum DedupResult {
+    /// skip
+    Skip,
+    /// write
+    Write,
+}
+impl DeDuper {
+    // emits the extra line saying 'last line repeated n times'
+    fn write(
+        w: &mut dyn Write,
+        encoder: &dyn Encode,
+        record: &Record,
+        n: i32,
+    ) -> Result<(), Box<dyn Error + Sync + Send>> {
+        if n == 1 {
+            encoder.encode(
+                w,
+                &Record::builder()
+                    .args(format_args!("last message repeated, suppressing dups"))
+                    .level(record.level())
+                    .target(record.target())
+                    .module_path_static(None)
+                    .file_static(None)
+                    .line(None)
+                    .build(),
+            )
+        } else {
+            encoder.encode(
+                w,
+                &Record::builder()
+                    .args(format_args!("last message repeated {} times", n))
+                    .level(record.level())
+                    .target(record.target())
+                    .module_path_static(None)
+                    .file_static(None)
+                    .line(None)
+                    .build(),
+            )
```

Comment:
instead of calling these encode fns in the if else branches can we just return the msg and call the encode after in one place? Seems like a lot of repeated lines

Reply:
tried that, format_args is amazingly hard to refactor

Reply:
I was thinking something like this

```rust
let args = if n == 1 {
  format_args!("last message repeated, suppressing dups")
} else {
  format_args!("last message repeated {} times", n)
};
encoder.encode(
    w,
    &Record::builder()
        .args(args)
        .level(record.level())
        .target(record.target())
        .module_path_static(None)
        .file_static(None)
        .line(None)
        .build(),
)
```